sonarts-core/src/rules/noIgnoredInitialValueRule.ts

```diff
@@ -0,0 +1,124 @@
+  }
+}
+
+class Walker extends tslint.ProgramAwareRuleWalker {
+  private lva: LiveVariableAnalyzer;
+
+  constructor(
+    sourceFile: ts.SourceFile,
+    options: tslint.IOptions,
+    program: ts.Program,
+    private readonly symbols: SymbolTable,
+  ) {
+    super(sourceFile, options, program);
+    this.lva = new LiveVariableAnalyzer(this.symbols);
+  }
+
+  protected visitNode(node: ts.Node): void {
+    if (is(node, ...FUNCTION_LIKE)) {
+      const functionLike = node as ts.FunctionLikeDeclaration;
+      const lvaReturn = this.lva.analyzeFunction(functionLike);
+      this.check(functionLike.body!, lvaReturn, ...functionLike.parameters);
+    } else if (is(node, ts.SyntaxKind.CatchClause)) {
+      const catchClause = node as ts.CatchClause;
+      const lvaReturn = this.lva.analyze(catchClause.block, catchClause.block.statements);
+      this.check(catchClause.block, lvaReturn, catchClause.variableDeclaration);
+    } else if (is(node, ts.SyntaxKind.ForInStatement, ts.SyntaxKind.ForOfStatement)) {
+      const iterationStatement = node as ts.IterationStatement;
+      const lvaReturn = this.lva.analyze(iterationStatement, [iterationStatement.statement]);
+      this.check(iterationStatement.statement, lvaReturn, (iterationStatement as ts.ForInStatement).initializer);
+    }
+
+    super.visitNode(node);
+  }
+
+  private check(root: ts.Node, lvaReturn?: LVAReturn,  ...nodesToCheck: ts.Node[]) {
+    if (!lvaReturn) {
+      return;
+    }
+    const { cfg, blockAvailableReads } = lvaReturn;
+    const symbolsLiveAtStart = blockAvailableReads.get(cfg.start)!;
+
+    nodesToCheck.forEach(parameter => {
+      descendants(parameter)
+        .filter(descendant => is(descendant, ts.SyntaxKind.Identifier))
+        .forEach(descendant => {
+          const identifier = descendant as ts.Identifier;
+          const symbol = this.getTypeChecker().getSymbolAtLocation(identifier);
+          if (
+            symbol &&
+            !symbolsLiveAtStart.has(symbol) &&
+            isVariableOrParameter(symbol) &&
+            this.onlyUsedLocallyToRoot(symbol, root) &&
+            this.symbols.allUsages(symbol).length > 1
+          ) {
+            this.addFailureAtNode(identifier, Rule.formatMessage(identifier));
+          }
+        });
+    });
+  }
+
+  private onlyUsedLocallyToRoot(symbol: ts.Symbol, root: ts.Node): boolean {
+    const boundaries = [
+      ...FUNCTION_LIKE,
+      ts.SyntaxKind.SourceFile,
+      ts.SyntaxKind.ClassDeclaration,
+      ts.SyntaxKind.ClassExpression,
+    ];
+    return this.symbols
+      .allUsages(symbol)
+      .every(
+        usage =>
+          usage.is(UsageFlag.DECLARATION) || ancestorsChain(usage.node, ...boundaries).includes(root),
+      );
+  }
+}
+
+function isVariableOrParameter(symbol: ts.Symbol) {
+  return symbol.flags === ts.SymbolFlags.BlockScopedVariable || symbol.flags === ts.SymbolFlags.FunctionScopedVariable;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') == should be === [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') == should be === [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)